products.js

```diff
@@ -0,0 +1,308 @@
+//
+//       //image title matches page title
+//       rule(type('images'), score(imageTitle)),
+//
+//       //image title contained in image src
+//       rule(type('images'), score(titleInSrc)),
+//
+//       //not to be confused with the thumbnail or logo
+//       rule(type('images'), score(keywords)),
+//
+//       //return image with max score
+//       rule(type('images').max(), out('product-image'))
+//
+//     );
+//
+//     function contentFnodes(doc) {
+//         return rules.against(doc).get('product-image');
+//     }
+//
+//     return contentFnodes;
+// }
+
+// function tunedTitleFnodes() {
+//     const rules = ruleset(
+//       //get all title tags in the inserted fixture
+//       rule(dom('div#fixture > title'), type('titleish')),
+//
+//       //return image with max score
+//       rule(type('titleish').max(), out('product-title'))
+//
+//     );
+//
+//     function contentFnodes(doc) {
+//         return rules.against(doc).get('product-title');
+//     }
+//
+//     return contentFnodes;
+// }
+
+// function tunedPriceFnodes() {
+//
+//     function hasDollarSign(fnode){
+//       if(fnode.element.childNodes[0] && fnode.element.childNodes[0].nodeValue && fnode.element.childNodes[0].nodeValue.includes('$')){
+//         return 2;
+//       }
+//       return 1;
+//     }
+//
+//     function tagHasGoodCss(fnode){
+//       if(fnode.element.id.match(/price/i) || fnode.element.classList.contains(/price/i)){
+//         return 2;
+//       }
+//       return 1;
+//     }
+//
+//     function notSavingsAmount(fnode){
+//       if(fnode.element.childNodes[0] && fnode.element.childNodes[0].nodeValue &&
+//         (fnode.element.childNodes[0].nodeValue.includes('-') || window.getComputedStyle(fnode.element).getPropertyValue('text-decoration') === 'line-through')){
+//         return 0;
+//       }
+//       return 1;
+//     }
+//
+//
+//     const rules = ruleset(
+//       //get all elements that could contain the price
+//       rule(dom('span, div, li'), type('priceish')),
+//
+//       //bonus if direct text (not children) contains a dollar sign
+//       rule(type('priceish'), score(hasDollarSign)),
+//
+//       //look for good css within tag
+//       rule(type('priceish'), score(tagHasGoodCss)),
+//
+//       //not to be confused with amount off (minus sign, crossed off)
+//       rule(type('priceish'), score(notSavingsAmount)),
+//
+//       //return image with max score
+//       rule(type('priceish').max(), out('product-price'))
+//
+//     );
+//
+//     function contentFnodes(doc) {
+//         return rules.against(doc).get('product-price');
+//     }
+//
+//     return contentFnodes;
+// }
+
+/**
+ * Maintain state as we compare a series of DOMs, reporting the percent
+ * difference at the end.
+ */
+class DiffStats {
+    constructor(contentFnodes, feature) {
+        this.lengthOfExpectedTexts = 0;
+        this.lengthOfDiffs = 0;
+        this.feature = feature;
+        if (feature === 'image') {
```

Comment:
I think I've seen this dispatch pattern before. :-) One way to factor it out is to, instead of just having `features = ['prices', 'floobs', 'flabs']`, let it be `features = [{name: 'prices', defaultRulesetMaker: tunedPriceFnodes}, ...]`. Then you don't need to sprinkle the knowledge that you're dealing with prices, images, and titles throughout the program. It makes things like DiffStats more general.